### ISSUE TITLE
Filling a gap lost link between fragments

### ISSUE THREAD

**Issue by bwindels** [bwindels/morpheusjs#15]
After filling a gap, the fill gap button disappeared while there were still more messages to be paginated in the gap (which I knew from having seen the message history on another client). This was a gap with more messages prior to it (due to a gappy sync). After refreshing, the messages above couldn't be loaded any more, the earliest message in the timeline being the earliest from filling the gap. Scrolling up didn't do anything and no fill gap button.

Might be related #10 , was on same device

**Comment by bwindels** [comment]
perhaps paginating beyond the room creation is related? this does cause an error, but things seem fine after reloading the room.

**Comment by bwindels** [comment]
Should not be related to #10, as that was completely separate from persistence. Here, the corruption is persisted, so likely something in GapWriter.

**Comment by bwindels** [comment]
Hmm, or would it? I can't remember if this bug still occurred after a refresh.

**Comment by bwindels** [comment]
I wonder if this can be about filling a gap with an overlapping response with fragments that are already linked up (e.g. gap of less than 10 events after limited sync).

Edit: **No, this is probably not the case as as we wrote in the description that there were more events to paginate.**

**Comment by bwindels** [comment]
are/should we be remove the gap entry when filling a gap?

**Comment by bwindels** [comment]
Could [linkedFragmentId](https://github.com/bwindels/brawl-chat/blob/0524f06722a5c7405cf6e9cd7103ea87c432adb5/src/matrix/room/timeline/persistence/GapWriter.js#L60) here be setting it for the wrong direction? Really need to repro this so we can inspect indexeddb.

Edit: unlikely, see above edit.

**Comment by bwindels** [comment]
Different idea: could it be a read-then-write error on the fragment between the SyncWriter and the GapWriter. So, with the timing just right, could the following happen (need to read up how idb would handle this):
 - GapWriter opens a transaction against the fragment store (this is sync in idb IIRC)
 - SyncWriter opens a transaction against the fragment store (this is sync in idb IIRC)
 - GapWriter reads the fragment
 - GapWriter adds previousId and writes the fragment
 - SyncWriter reads the fragment
 - SyncWriter adds nextId and writes the fragment
 - SyncWriter commits transaction (nextId will now be visible to other readers)
 - GapWriter commits transaction (previousId will now be visible to other readers)

I remember I wanted to make opening a transaction async as well to prevent this kind of error but never fully researched if this was a problem. Important caveat here is that both transactions are `readwrite`.

**Comment by bwindels** [comment]
https://w3c.github.io/IndexedDB/#transaction-scheduling implies that the above is not possible, assuming that Edge Mobile 15 is spec-compliant.

**Comment by bwindels** [comment]
Marking this to be closed in the future as it happened only once, on a device I am not using anymore.

**Comment by bwindels** [comment]
I guess we could try to build a little prototype to try the above indexeddb theory?

**Comment by bwindels** [comment]
Made a prototype, works well on lumia, so don't think the idb theory pans out.

**Comment by bwindels** [comment]
Just hit (something like) this again with 0.0.19!
While back-paginating in `Matrix & Friend`, I hit this:

`Could not load next messages: Next fragment 19 doesn't point back to 18`

![image](https://user-images.githubusercontent.com/274386/77820428-3c980a80-70da-11ea-81c7-9a24c7fb809f.png)


Fragments as pulled straight out of indexeddb:
```
roomId: "!WCHwNGGsrVwPeRtAnq:matrix.org"
id: 18
previousId: 17
nextId: 19
previousToken: null
nextToken: null
key: "!WCHwNGGsrVwPeRtAnq:matrix.org|00000012"

roomId: "!WCHwNGGsrVwPeRtAnq:matrix.org"
id: 19
previousId: 18
nextId: null
previousToken: "t116230-1314852630_757269739_1037333_466964661_301544009_1144442_46059099_22957226_86880"
nextToken: null
key: "!WCHwNGGsrVwPeRtAnq:matrix.org|00000013"
```

Stack trace from error:
```
brawl.js:4004 timeline.fillGap(): Next fragment 19 doesn't point back to 18:
Error: Next fragment 19 doesn't point back to 18
    at findForwardSiblingFragments (https://bwindels.github.io/brawl/brawl.js:2866:23)
    at createIslands (https://bwindels.github.io/brawl/brawl.js:2888:34)
    at FragmentIdComparer.rebuild (https://bwindels.github.io/brawl/brawl.js:2958:29)
    at FragmentIdComparer.add (https://bwindels.github.io/brawl/brawl.js:2971:18)
    at GapWriter._updateFragments (https://bwindels.github.io/brawl/brawl.js:1936:42)
    at GapWriter.writeFragmentFill (https://bwindels.github.io/brawl/brawl.js:1977:24)
    at async Room.fillGap (https://bwindels.github.io/brawl/brawl.js:3248:30)
    at async GapTile.fill (https://bwindels.github.io/brawl/brawl.js:4002:21)
fill @ brawl.js:4004
brawl.js:566 Uncaught (in promise) Error: add on undefined failed: (name: AbortError) (code: AbortError) The transaction was aborted, so the request cannot be fulfilled.
    at Store.add (brawl.js:566)
brawl.js:557 Uncaught (in promise) Error: put on undefined failed: (name: AbortError) (code: AbortError) The transaction was aborted, so the request cannot be fulfilled.
    at Store.put (brawl.js:557)
```

After refreshing, I was able to fill the gap. So this is likely a separate issue?

**Comment by bwindels** [comment]
https://matrix.org/docs/spec/client_server/latest#syncing actually has a warning about duplicate events, which is exactly what we reported at https://github.com/matrix-org/synapse/issues/7164, so specced behaviour, not a server bug.
